pkg/queryprojection/analyzer_test.go

```diff
@@ -0,0 +1,201 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package queryprojection
+
+import (
+	"sort"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestAnalyzeQuery(t *testing.T) {
+	analyzer := NewQueryAnalyzer()
+
+	for _, tc := range []struct {
+		name             string
+		expression       string
+		labels           []string
+		expectedBy       bool
+		expectedGrouping bool
+		expectedLabels   []string
+	}{
+		{
+			name:       "vector selector, not projectable",
+			expression: "up",
+		},
+		{
+			name:       "matrix selector, not projectable",
+			expression: "up[5m]",
+		},
+		{
+			name:       "contains label_replace, not projectable",
+			expression: `label_replace(sum(up), "dst_label", "$1", "src_label", "re")`,
+		},
+		{
+			name:       "binary expression",
+			expression: `http_requests_total{code="400"} / http_requests_total`,
+		},
+		{
+			name:       "binary expression with constant",
+			expression: `http_requests_total{code="400"} / 4`,
+		},
+		{
+			name:             "binary expression with empty vector matching",
+			expression:       `http_requests_total{code="400"} / on () http_requests_total`,
+			expectedBy:       true,
+			expectedGrouping: true,
+		},
+		{
+			name:             "binary expression with vector matching",
+			expression:       `http_requests_total{code="400"} / on (pod) http_requests_total`,
+			expectedBy:       true,
+			expectedGrouping: true,
+			expectedLabels:   []string{"pod"},
```

Comment:
I don't think this is correct, because we need all labels for the end-result. I don't think we can apply a projection unless there is an aggregation in the expression.

Reply:
Where do you use other labels?
I tested [here](https://prometheus.demo.do.prometheus.io/graph?g0.expr=prometheus_http_requests_total%7Bcode%3D%22200%22%2C%20handler%3D%22%2F-%2Fhealthy%22%7D%20%2F%20on%20(instance)%20sum(prometheus_http_requests_total)%20by%20(instance)&g0.tab=1&g0.stacked=0&g0.range_input=1h) and I can see the end result only contains the matching labels. You mena the promql engine needs other labels for something else?

Reply:
Oh wow, I didn't know promql worked like this 😄 Ok this makes sense then 👍